### PR TITLE
Standard entity metadata to help measure Scaffolder ROI 

### PR DESCRIPTION
**Pull request by iamEAP**
## Need

Backstage takes time and effort to stand up and maintain, but the "create" use-case of Backstage (as fulfilled by the Scaffolder plugin) provides a straightforward way to quantify the value of Backstage usage: time savings!

Given a software template definition, it's impossible to tell _how much_ time would be saved by leveraging the template (vs. not having a template at all).  We therefore need some way to associate time savings for a given template.

Assuming we have a that in place, we also still need some way to understand the _magnitude_ of the savings (e.g. how many times was a template run?)

## Proposal

Introduce two new annotations to encode the above data points:

- `backstage.io/time-saved`: Representing the time that someone would save by using the template (compared to reproducing what the template encodes by hand).  Formatted as an ISO duration. Intended to be used on entities of kind `Template`.
- `backstage.io/source-template`: Representing the source template entity that a given entity was created from.

### Uses

- The `backstage.io/time-saved` annotation is converted to `minutes` and added as the `value` on the scaffolder's `create` analytics event (summing this can yield "time saved" in an analytics tool). This therefore proposes that the primary measure associated with `create` events is the amount of time saved.
- The `backstage.io/source-template` annotation is set automatically on the `catalog:write` scaffolder action (but can also manually be set in more manual template use-cases). This value is exposed in the `<AboutCard>` as a "Create Something Similar" button (see below)
  
  Create Something Similar

Even without an analytics set up, with the above, we could query the Catalog for an aggregation (count) of entities that have a `backstage.io/source-template` value.  We could then multiply the counts by the corresponding `backstage.io/time-saved` value in order to get a rough sense of overall time savings yielded by leveraging the Scaffolder.

## Alternatives

- **Don't introduce a standard**: We could certainly go about solving this in myriad ways among Backstage adopters. It would be unfortunate not to have at least a basic standard for this, given how natural it is to measure ROI in this way.

- **Take inspiration from prior art**: There exists the [TimeSaver plugin](https://github.com/tduniec/backstage-timesaver-plugin?tab=readme-ov-file#generate-statistics), which includes an alternative schema for storing time savings metadata on Template entities. This plugin includes the ability to breakdown time savings by category. Perhaps this is better suited as a custom superset of the proposed standard in this PRFC (where one could imagine collapsing this plugin's schema into the proposed `time-saved` annotation), with this plugin providing extra functionality on top of the standard.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: docs/features/software-catalog/well-known-annotations.md
Comment (HHogg) [inline docs/features/software-catalog/well-known-annotations.md:155]: 🔥

File: plugins/catalog/src/components/AboutCard/AboutCard.tsx
Comment (HHogg) [inline plugins/catalog/src/components/AboutCard/AboutCard.tsx:144]: Should this also do a check to see if the template still exists and can be created from it? To handle cases where templates are removed.

Comment (iamEAP) [reply plugins/catalog/src/components/AboutCard/AboutCard.tsx:144]: It's a good thought, but it would require an extra lookup when the card gets rendered.

File: docs/features/software-catalog/well-known-annotations.md
Comment (HHogg) [reply docs/features/software-catalog/well-known-annotations.md:155]: How does this get added? Can it be done automatically by scaffolder?

Comment (iamEAP) [reply docs/features/software-catalog/well-known-annotations.md:155]: Only viable path in my mind is for it to be encoded in the underlying template by the owner of the Template (e.g. in cases where the template includes a `catalog-info.yaml` file that gets registered).  Can't really think of a viable way for those whose entities are ingested by non-declarative means.

Comment (iamEAP) [reply docs/features/software-catalog/well-known-annotations.md:155]: @benjdlambert sayz: maybe the catalog write action could be a spot to help automate this.

File: plugins/catalog/src/components/AboutCard/AboutCard.tsx
Comment (iamEAP) [reply plugins/catalog/src/components/AboutCard/AboutCard.tsx:144]: I suppose it might also be good to do this anyway, from an authz perspective.

**Comment by iamEAP** [comment]
> _Feedback based on conversation at Scaffolder SIG:_
>
> - _Possible opportunities to pass the the `time-saved` duration via prometheus_

Alright, after taking a look at some of the prometheus observability setup, I'm not sure it makes sense to try and add `time-saved`, there.  Aggregating metrics that aren't event counts can be a little counter-unintuitive.  ...I also continue to be skeptical about mixing o11n and analytics data: they have different audiences who are trying to solve for different needs.  Calculating ROI is definitely more in line with the analytics side of the house than o11n.

------

Given that, I think this is ready for another set of eyes.

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx
Comment (HHogg) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:154]: Is it ok that the property is called `value`?  Should this sit under `attributes` or `context`?

File: plugins/scaffolder-react/src/next/hooks/useTemplateTimeSaved.ts
Comment (HHogg) [inline plugins/scaffolder-react/src/next/hooks/useTemplateTimeSaved.ts:45]: Not that there's anything wrong with the calculation here but could you use luxons `as('minutes')` here to save the millis calculation. 

```ts
Duration.fromISO(timeSaved).as('minutes') // NaN for an invalid duration
```

Comment (iamEAP) [reply plugins/scaffolder-react/src/next/hooks/useTemplateTimeSaved.ts:45]: I was looking for this in the API docs!  "Surely there must be a way to convert a duration to an arbitrary amount of a given unit."  But I only saw `toMillis()`.

Thanks. Updating!

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx
Comment (iamEAP) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:154]: Yeah, it's [a core part](https://backstage.io/docs/plugins/analytics/#providing-extra-attributes) of an analytics event.  Every event can have one numeric value associated with it, typically representing the "primary" metric one might want to know for a given action.  Existing uses of this include...

- `search` - The number of results returned.
- `discover` - The "rank" of the search result that was clicked on.

This is proposing that the "primary" metric to capture on `create` events is the "amount of time saved."

Comment (HHogg) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:154]: I'm kinda worried this could get misunderstood for something else, e.g. "Time it took to run" or "Index of the run" or  ..etc. Is it ok that we use the _only_ value property to represent this?

**Review by nuritizra — APPROVED** [review]

File: plugins/catalog/src/components/AboutCard/AboutCard.tsx
Comment (Rugvip) [inline plugins/catalog/src/components/AboutCard/AboutCard.tsx:253]: Thinking that using the `scaffolder` icon from the app icons might be a good option too?

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx
Comment (Rugvip) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:154]: Should we subtract the amount of time it took to fill out the form and wait for the templating task to run? 😏

File: docs/features/software-catalog/descriptor-format.md
Comment (Rugvip) [inline docs/features/software-catalog/descriptor-format.md:743]: Best have a few examples here of what that format looks like

**Review by Rugvip — APPROVED** [review]
Interesting stuff! 😁 

No major blockers but a few nits. Curious what thoughts on this are from @backstage/scaffolder-maintainers, and is this something you've discussed internally @acierto?

**Comment by iamEAP** [comment]
> _Similar feature can be added to backstage by installing https://github.com/tduniec/backstage-timesaver-plugin_

Thanks @tduniec!  Definitely aware, and tried to address in the PR description.

I'd be happy to contribute a Catalog processor in your monorepo that looked for the `substitute` metadata key on templates and rolled up the various values into a `time-saved` annotation, so that the users of the `time-saver` plugin could automatically benefit from this proposed standard too.

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx
Comment (benjdlambert) [inline plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:79]: It's a little unfortunate to add this to the `Stepper` props.

Is there a better way that we can refactor this instead rather than leaking out this detail into the props?

Wondering if we can move the tracking entirely to a parent component? Or provide something more generic like `trackingContext` or something as a prop?

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Just a small thing that I want to get through - this is a public component, and a little unfortunate to be leaking details. Feels a little strange to me.

File: plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx
Comment (iamEAP) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:79]: Yeah lemme see what I can do.

Comment (iamEAP) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:79]: > _Wondering if we can move the tracking entirely to a parent component?_

So my worry with moving the tracking out of the stepper itself is that any custom usages of the stepper (outside the context of `<Workflow>`, for example) would suddenly lose tracking.  ...Which isn't ideal.

Happy to add a `trackingContext` prop (and deprecate `templateName`, which is only used in the tracking code) though.

Comment (benjdlambert) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx:79]: Yeah something like that, If you can find a better name also please go ahead, but thinking it's a little cleaner than adding more props for now.

:tada:

Comment (iamEAP) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx]: Per discussion offline: given the `<Stepper>` probably shouldn't be used outside the context of a `<Workflow>` anyway, it makes sense to lift the event capture up there and then not have to deal with passing props down at all.

Doing so may even fix a bug where the `create` event was possibly being fired in the template editor form.

Comment (benjdlambert) [reply plugins/scaffolder-react/src/next/components/Stepper/Stepper.tsx]: Nice!

**Review by benjdlambert — APPROVED** [review]
Nice - let's go with this and see where we land :tada:

**Pull request merged**
